ssl/t1_lib.c

```diff
@@ -1137,25 +1137,180 @@ static const uint16_t tls_default_sigalg[] = {
+        /*
+         * We have successfully fetched the algorithm - however if the provider
+         * doesn't match this one then we ignore it.
+         *
+         * Note: We're cheating a little here. Technically if the same algorithm
+         * is available from more than one provider then it is undefined which
+         * implementation you will get back. Theoretically this could be
+         * different every time...we assume here that you'll always get the
+         * same one back if you repeat the exact same fetch. Is this a reasonable
+         * assumption to make (in which case perhaps we should document this
+         * behaviour)?
+         */
+        if (EVP_KEYMGMT_get0_provider(keymgmt) == provider) {
+            /* We have a match - so we will use this sigalg */
+           (*num_from_provider)++;
+           tmp = NULL;
+        }
+        EVP_KEYMGMT_free(keymgmt);
+    }
+ err:
+    if (tmp) {
+        OPENSSL_free(tmp->name);
+        memset(tmp, 0, sizeof(SIGALG_LOOKUP));
+    }
+    return ret;
+}
+
+static int discover_provider_sigalgs(OSSL_PROVIDER *provider, void *vctx)
+{
+    struct provider_sigalg_data_st *psd = vctx;
+
+    psd->provider = provider;
+    OSSL_PROVIDER_get_capabilities(provider, "TLS-SIGALG",
+                                   add_provider_sigalgs, psd);
+    /* If a provider returns 0 it means it does not support this capability.
+     * We should continue on looking at other providers so return 1 no matter
+     * what the function above returns.
+     */
+    return 1;
+}
+
 int ssl_setup_sig_algs(SSL_CTX *ctx)
 {
     size_t i;
     const SIGALG_LOOKUP *lu;
-    SIGALG_LOOKUP *cache
-        = OPENSSL_malloc(sizeof(*lu) * OSSL_NELEM(sigalg_lookup_tbl));
+    SIGALG_LOOKUP *from_provider = NULL;
+    size_t num_from_provider = 0;
+    struct provider_sigalg_data_st psd;
+    SIGALG_LOOKUP *cache = NULL;
     EVP_PKEY *tmpkey = EVP_PKEY_new();
     int ret = 0;
 
-    if (cache == NULL || tmpkey == NULL)
+    if (!tmpkey)
+        return 0;
+
+    psd.ctx = ctx;
+    psd.from_provider = &from_provider;
+    psd.num_from_provider = &num_from_provider;
+    psd.provider = NULL; /* Set in the discover() function. */
+    if (!OSSL_PROVIDER_do_all(ctx->libctx, discover_provider_sigalgs, &psd))
+        goto err;
+
+    cache = OPENSSL_zalloc(sizeof(*lu) * (OSSL_NELEM(sigalg_lookup_tbl) + num_from_provider));
+    if (!cache)
         goto err;
 
     ERR_set_mark();
     for (i = 0, lu = sigalg_lookup_tbl;
-         i < OSSL_NELEM(sigalg_lookup_tbl); lu++, i++) {
+         i < OSSL_NELEM(sigalg_lookup_tbl) + num_from_provider; lu++, i++) {
         EVP_PKEY_CTX *pctx;
 
-        cache[i] = *lu;
-
+        if (i < OSSL_NELEM(sigalg_lookup_tbl)) {
+            cache[i] = *lu;
+            /* We need to do this because the name is malloc()'d for the ones
+             * from the provider.
+             */
+            if (cache[i].name) {
+                cache[i].name = strdup(lu->name);
```

Comment:
Use `OPENSSL_strdup` instead.
An alternative would be to have some kind of indicator in `SIGALG_LOOKUP` to show whether the name was malloc'd or not - and only free those that were actually malloc'd. Then you could copy the whole `sigalg_lookup_tbl` in a single `memcpy`.

Reply:
> An alternative would be to have some kind of indicator in `SIGALG_LOOKUP` to show whether the name was malloc'd or not - and only free  those that were actually malloc'd. 

That puts the complexity during freeing which I would prefer to avoid.  However, this is a stylistic decision so I am open to the suggestion.  Does it better match OpenSSL's style to do so?

Reply:
We probably have both approaches in OpenSSL so from a stylistic perspective either could work. It just seems more efficient to me to copy everything all in one go. It's not really that much additional complexity in the freeing.